### DISCUSSION TITLE
[A-L] (7 luglio 2023) Esercizio 3

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#77] [Esami A-L]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/d282cfb2-b08e-467d-a8ef-9c3daa77aeba)

**Comment by Frisbon** [comment]
### 2 Cammini (viola e giallo) che possono resituire YES
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/79e587ef-0fdd-4470-8b29-572a2bb9de9a)

**Comment by Frisbon** [comment]
``` python
def es3(G, x, y):
    VIS=[0,...,0]; VIS[x] = 1
    return DFS_ric(VIS,x,y,False,False,False,False)

def DFS_ric(VIS,x,target,r,g,b,result):
    
    VIS[x] = 1
    if x is RED: r = True
    elif x is BLUE: b = True
    else: g = True
    
    for y in x: 
        if y == target:
            if y is RED: r = True
            elif y is BLUE: b = True
            else: g = True
            
            return (r and g and b) #ritorna se tutti i colori sono stati trovati 
        
        if VIS[y] == 0:
            result = result or DFS_ric(VIS,y,target,r,g,b)
    
    return result   
```

**Comment by cornflexxx** [comment]
```
RGB_path(G, x, y):

	R ->(1,0,0)
	G ->(0,1,0)
	B ->(0,0,1)

	T := [(0,0,0)....,(0,0,0)]
	RGB_dfs(G,T,x,y)
	if T[x] == (1,1,1):
		return "YES"
	else: 
		return "NO"
	

RGB_dfs(G,T,v,y):
	if v == y:
		T[v] = C(v)
		return
	for u in v.uscenti:
	     if T[u] == (0,0,0):
		 RBG_dfs(G,T,u,y)
	     T[v] = T[u] | T[v]
	if T[v] != (0,0,0):
		T[v] = T[v] | C(v)
		
```

**Reply by Elia-Belli** [reply]
Mi sfugge qualcosa, nel grafo potrebbe esistere un cammino da $x\to y$ con dei colori assenti e il DFS può arbitrariamente visitare questo anziché il cammino colorato (nel caso esistesse).

Inoltre dal codice mi sembra che $T[x]$ sia uguale a $(1,1,1)$ se tra i suoi discendenti c'è almeno un nodo di ogni colore, e non necessariamente nel cammino da $x \to y$ : il DFS potrebbe non mirare direttamente ad $y$ e quindi il caso base `if v == y: ... return` è inutile.

**Comment by daromacarlo** [comment]
Strategia di programmazione dinamica. Costo = O(nm).
Quando una certa cella della tabella T è 0 significa che r = 0, g = 0, b = 0
                                      //                     è 1 significa che r = 0, g = 0, b = 1
                                      //                     è 2 significa che r = 0, g = 1, b = 0
                                      //                     è 3 significa che r = 0, g = 1, b = 0 [ecc...]

![Immagine 2025-01-27 175306](https://github.com/user-attachments/assets/f646fe1d-fa93-47b1-ac15-dbf473d1085d)
![Immagine 2025-01-27 175333](https://github.com/user-attachments/assets/4c3c227f-ed17-42d5-a467-a848e480a56e)

**Comment by cek753** [comment]
<img width="873" height="1125" alt="image" src="https://github.com/user-attachments/assets/9b65a3b9-919f-4cb9-a20a-88b7e0f3e5b6" />
